homeassistant/components/vallox/__init__.py

```diff
@@ -91,110 +89,111 @@
             return None
 
         if not isinstance(value, (str, int, float)):
-            raise TypeError(
-                f"Return value of metric {metric_key} has unexpected type {type(value)}"
+            _LOGGER.error(
+                "Return value of metric %s has unexpected type %s",
+                metric_key,
+                type(value),
             )
+            return None
 
         return value
 
-    def get_profile(self) -> VALLOX_PROFILE:
-        """Return cached profile value."""
-        _LOGGER.debug("Returning profile")
 
-        if not self._valid:
-            raise OSError("Device state out of sync.")
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the client and boot the platforms."""
+    conf = config[DOMAIN]
+    host = conf.get(CONF_HOST)
+    name = conf.get(CONF_NAME)
 
-        return self._profile
+    client = Vallox(host)
 
-    async def async_update(self, time: datetime | None = None) -> None:
+    async def async_update_data() -> ValloxState:
         """Fetch state update."""
         _LOGGER.debug("Updating Vallox state cache")
 
         try:
-            self._metric_cache = await self._client.fetch_metrics()
-            self._profile = await self._client.get_profile()
+            metric_cache = await client.fetch_metrics()
+            profile = await client.get_profile()
 
         except (OSError, ValloxApiException) as err:
-            self._valid = False
-            _LOGGER.error("Error during state cache update: %s", err)
-            return
+            raise UpdateFailed("Error during state cache update") from err
+
+        return ValloxState(metric_cache, profile)
+
+    coordinator = DataUpdateCoordinator[ValloxState](
+        hass,
+        _LOGGER,
+        name=f"{name} DataUpdateCoordinator",
+        update_interval=STATE_SCAN_INTERVAL,
+        update_method=async_update_data,
+    )
 
-        self._valid = True
-        async_dispatcher_send(self._hass, SIGNAL_VALLOX_STATE_UPDATE)
+    # The vallox hardware expects quite strict timings for websocket requests. Timings that machines
+    # with less processing power, like a Raspberry Pi, cannot live up to during the busy start phase
+    # of Home Asssistant. Hence, "async_add_entities()" for any platform code will be called with
+    # "update_before_add=False" to intentionally delay the first request, increasing chance that it
+    # is issued only when the machine is less busy again.
+    #
+    # For the same reason, don't call "coordinator.async_refresh()" here. Instead, set
+    # "coordinator.last_update_success" to "False", so that Home Assistant does not try to update
+    # the state of entities whose code assumes that "coordinator.data" is never "None".
+    #
+    # In summary, this will cause the first websocket requests to the Vallox client to go out after
+    # one "coordinator.update_interval" has passed.
+    coordinator.last_update_success = False
```

Comment:
Does it make sense to delay starting all of this until after the STARTED event?

Reply:
I can try. Do you have a pointer how to properly do this?